crates/doctor/src/rustlang/target.rs

```diff
@@ -0,0 +1,193 @@
+                .unwrap_or_default()
+        });
+
+        if uses_rust {
+            // does any component contain a build command with `cargo` as the program?
+            // if so, run rustup target list --installed and:
+            // - if rustup is not present, return RustNotInstalled
+            // - if rustup is present but the list does not contain wasm32-wasi, return WasmTargetNotInstalled
+            // - if the list does contain wasm32-wasi, return an empty list
+            // NOTE: this does not currently check against the Rust SDK MSRV - that could
+            // be a future enhancement or separate diagnosis, but at least the Rust compiler
+            // should give a clear error for that!
+            let target_list_output = tokio::process::Command::new("rustup")
+                .args(["target", "list", "--installed"])
+                .output()
+                .await;
+            let diagnosis = match target_list_output {
+                Err(e) => {
+                    if e.kind() == std::io::ErrorKind::NotFound {
+                        vec![TargetDiagnosis::RustNotInstalled]
+                    } else {
+                        anyhow::bail!(e)
+                    }
+                }
+                Ok(output) => {
+                    let stdout = String::from_utf8_lossy(&output.stdout);
+                    if stdout.lines().any(|line| line == "wasm32-wasi") {
+                        vec![]
+                    } else {
+                        vec![TargetDiagnosis::WasmTargetNotInstalled]
+                    }
+                }
+            };
+            Ok(diagnosis)
+        } else {
+            Ok(vec![])
+        }
+    }
+}
+
+/// TargetDiagnosis represents a problem with the Rust target.
+#[derive(Debug)]
+pub enum TargetDiagnosis {
+    /// Rust is not installed
+    RustNotInstalled,
+    /// The Rust wasm32-wasi target is not installed
+    WasmTargetNotInstalled,
+}
+
+impl Diagnosis for TargetDiagnosis {
+    fn description(&self) -> String {
+        match self {
+            Self::RustNotInstalled => "The Rust compiler isn't installed".into(),
+            Self::WasmTargetNotInstalled => {
+                "The required Rust target 'wasm32-wasi' isn't installed".into()
+            }
+        }
+    }
+
+    fn is_critical(&self) -> bool {
+        true
+    }
+
+    fn treatment(&self) -> Option<&dyn Treatment> {
+        Some(self)
+    }
+}
+
+#[async_trait]
+impl RustTreatment for TargetDiagnosis {
+    fn summary(&self) -> String {
+        match self {
+            Self::RustNotInstalled => "Install the Rust compiler and the wasm32-wasi target",
+            Self::WasmTargetNotInstalled => "Install the Rust wasm32-wasi target",
+        }
+        .into()
+    }
+
+    async fn treat_rust(&self) -> Result<()> {
+        match self {
+            Self::RustNotInstalled => {
+                install_rust().await?;
+                install_wasi_target()?;
```

Comment:
On that note, I wonder if we should try to detect that rustup has been installed but the user still needs to restart their terminal. Just checking for the existence of `~/.rustup/` (or the Windows equivalent) may be a good-enough heuristic and could help reduce annoyance if people don't see the output of the rust install step (which tells you to restart your terminal).

Reply:
Thanks, good catch and thanks for the tip re passing arguments.

Windows runs `rustup-init` directly (we manage the download) so I should be able to simply pass the `--target` option.  WHAT COULD GO WRONG

Reply:
Okay, that seems to work.  You are right, though, about the risk of the PATH not being updated.  This is particularly vexing because `spin doctor` wants to charge on and have you build your application immediately, which of course fails with "Rust not installed."  Huh?  You told me you installed it.  Call yourself a doctor?

We could introduce a notion of a stopping treatment, after which further diagnoses should be suppressed (or are listed but indicated as "re-run to treat").  It might be a bit of a blunt instrument but I'm not sure how else to manage that transition.

Reply:
I can think of a few "magical" solutions like updating `PATH` for the current process but those generally seem too surprising.

Thinking about the implementation of "stopping treatment", given that this is (hopefully) an exceptional kind of treatment I think it would be fine to return an error like `UserInterventionRequired { message: String, continue_treatments: bool }`.

Reply:
Done